File: packages/pythonpkg/src/pythonpkg/prime.py
import math
import time
from concurrent.futures import ThreadPoolExecutor, as_completed
from typing import List


def naive_prime_sieve(n: int) -> List[int]:
    """朴素试除法：检查每个数是否能被小于它的数整除"""
    primes = []
    for i in range(2, n + 1):
        is_prime = True
        for j in range(2, i):
            if i % j == 0:
                is_prime = False
                break
        if is_prime:
            primes.append(i)
    return primes


def optimized_naive_prime_sieve(n: int) -> List[int]:
    """优化试除法：只检查到 sqrt(i)"""
    primes = []
    for i in range(2, n + 1):
        if i == 2:
            primes.append(i)
            continue
        if i % 2 == 0:
            continue
        is_prime = True
        for j in range(2, int(math.isqrt(i)) + 1):
            if i % j == 0:
                is_prime = False
                break
        if is_prime:
            primes.append(i)
    return primes


def _is_prime_single(i: int) -> int:
    """辅助函数：判断单个数是否为质数（用于多线程）"""
    if i < 2:
        return None
    if i == 2:
        return i
    if i % 2 == 0:
        return None
    for j in range(3, int(math.isqrt(i)) + 1, 2):
        if i % j == 0:
            return None
    return i


def omp_prime_sieve(n: int) -> List[int]:
    """模拟 OpenMP 多线程试除法"""
    primes = []
    with ThreadPoolExecutor() as executor:
        futures = [executor.submit(_is_prime_single, i) for i in range(2, n + 1)]
        for future in as_completed(futures):
            result = future.result()
            if result is not None:
                primes.append(result)
    primes.sort()  # 保证顺序
    return primes


def eratosthenes_sieve(n: int) -> List[int]:
    """埃拉托斯特尼筛法"""
    if n < 2:
        return []
    is_prime = [True] * (n + 1)
    is_prime[0] = is_prime[1] = False
    for i in range(2, int(math.isqrt(n)) + 1):
        if is_prime[i]:
            for j in range(i * i, n + 1, i):
                is_prime[j] = False
    return [i for i in range(2, n + 1) if is_prime[i]]


def linear_sieve(n: int) -> List[int]:
    """线性筛法（欧拉筛）"""
    if n < 2:
        return []
    min_prime = [0] * (n + 1)  # min_prime[i] 表示 i 的最小质因数
    primes = []
    for i in range(2, n + 1):
        if min_prime[i] == 0:
            min_prime[i] = i
            primes.append(i)
        j = 0
        while j < len(primes) and i * primes[j] <= n:
            min_prime[i * primes[j]] = primes[j]
            if i % primes[j] == 0:
                break
            j += 1
    return primes

File: packages/cpkg/src/cpkg/prime.c
// prime.c
#include "prime.h"
#include <stdlib.h>
#include <math.h>
#include <omp.h>

// 初始化动态数组
void long_vector_init(LongVector* vec) {
    vec->size = 0;
    vec->capacity = 16;
    vec->data = (long*)malloc(vec->capacity * sizeof(long));
    if (!vec->data) exit(1); // 简单错误处理
}

// 扩容
void long_vector_reserve(LongVector* vec, int new_capacity) {
    if (new_capacity > vec->capacity) {
        vec->data = (long*)realloc(vec->data, new_capacity * sizeof(long));
        if (!vec->data) exit(1);
        vec->capacity = new_capacity;
    }
}

// 添加元素
void long_vector_push(LongVector* vec, long value) {
    if (vec->size >= vec->capacity) {
        long_vector_reserve(vec, vec->capacity * 2);
    }
    vec->data[vec->size++] = value;
}

// 释放
void long_vector_free(LongVector* vec) {
    free(vec->data);
    vec->data = NULL;
    vec->size = vec->capacity = 0;
}

// 1. 暴力法：检查每个数是否被小于它的数整除
void naive_prime_sieve(long n, LongVector* primes) {
    for (long i = 2; i <= n; ++i) {
        bool is_prime = true;
        for (long j = 2; j < i; ++j) {
            if (i % j == 0) {
                is_prime = false;
                break;
            }
        }
        if (is_prime) {
            long_vector_push(primes, i);
        }
    }
}

// 2. 优化暴力法：只检查到 sqrt(i)
void optimized_naive_prime_sieve(long n, LongVector* primes) {
    for (long i = 2; i <= n; ++i) {
        bool is_prime = true;
        long limit = (long)sqrt(i);
        for (long j = 2; j <= limit; ++j) {
            if (i % j == 0) {
                is_prime = false;
                break;
            }
        }
        if (is_prime) {
            long_vector_push(primes, i);
        }
    }
}

// 3. OpenMP 并行版本
void omp_prime_sieve(long n, LongVector* primes) {
    #pragma omp parallel
    {
        LongVector thread_primes;
        long_vector_init(&thread_primes);

        long int i,k;
        #pragma omp for
        for (i = 2; i <= n; ++i) {
            bool is_prime = true;
            long limit = (long)sqrt(i);
            for (long j = 2; j <= limit; ++j) {
                if (i % j == 0) {
                    is_prime = false;
                    break;
                }
            }
            if (is_prime) {
                long_vector_push(&thread_primes, i);
            }
        }

        // 合并结果
        #pragma omp critical
        {
            for (k = 0; k < thread_primes.size; ++k) {
                long_vector_push(primes, thread_primes.data[k]);
            }
        }

        long_vector_free(&thread_primes);
    }
}

// 4. 埃拉托斯特尼筛法
void eratosthenes_sieve(long n, LongVector* primes) {
    bool* is_prime = (bool*)calloc(n + 1, sizeof(bool));
    for (long i = 2; i <= n; ++i) is_prime[i] = true;

    for (long i = 2; i * i <= n; ++i) {
        if (is_prime[i]) {
            for (long j = i * i; j <= n; j += i) {
                is_prime[j] = false;
            }
        }
    }

    for (long i = 2; i <= n; ++i) {
        if (is_prime[i]) {
            long_vector_push(primes, i);
        }
    }

    free(is_prime);
}

// 5. 线性筛（欧拉筛）
void linear_sieve(long n, LongVector* primes) {
    long* min_prime = (long*)calloc(n + 1, sizeof(long)); // 记录最小质因数

    for (long i = 2; i <= n; ++i) {
        if (min_prime[i] == 0) {
            min_prime[i] = i;
            long_vector_push(primes, i);
        }
        // 筛去所有已知质数与 i 的乘积
        for (int j = 0; j < primes->size; ++j) {
            long p = primes->data[j];
            long next = i * p;
            if (next > n) break;
            min_prime[next] = p;
            if (i % p == 0) break; // 保证每个数只被最小质因数筛一次
        }
    }

    free(min_prime);
}